shell/platform/linux/fl_engine.cc

```diff
@@ -125,6 +282,106 @@ gboolean fl_engine_start(FlEngine* self, GError** error) {
   return TRUE;
 }
 
+G_MODULE_EXPORT void fl_engine_set_platform_message_handler(
+    FlEngine* self,
+    const gchar* channel,
+    FlEnginePlatformMessageCallback callback,
+    gpointer user_data) {
+  g_return_if_fail(FL_IS_ENGINE(self));
+  g_return_if_fail(channel != nullptr);
+  g_return_if_fail(callback != nullptr);
+
+  PlatformMessageHandler* handler =
+      platform_message_handler_new(callback, user_data);
+  g_hash_table_replace(self->platform_message_handlers, g_strdup(channel),
+                       handler);
+}
+
+G_MODULE_EXPORT gboolean
+fl_engine_send_platform_message_response(FlEngine* self,
+                                         FlEnginePlatformMessageHandle* handle,
+                                         GBytes* response,
+                                         GError** error) {
+  g_return_val_if_fail(FL_IS_ENGINE(self), FALSE);
+  g_return_val_if_fail(handle != nullptr, FALSE);
+
+  gsize data_length = 0;
+  const uint8_t* data = nullptr;
+  if (response != nullptr)
+    data =
+        static_cast<const uint8_t*>(g_bytes_get_data(response, &data_length));
+  FlutterEngineResult result = FlutterEngineSendPlatformMessageResponse(
+      self->engine, handle->message->response_handle, data, data_length);
+  g_ptr_array_remove(self->message_handles, handle);
+
+  if (result != kSuccess) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Failed to send platorm message response");
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+G_MODULE_EXPORT void fl_engine_send_platform_message(
```

Comment:
You're not going to want these APIs to be on the engine:
- Plugins need access to messaging, but shouldn't get access to the whole engine API.
- We recently had to fix retain cycle leaks in some of the retain-count-based embeddings due the fact that the object given to plugins was actually the engine under the hood.

You should make a messenger interface like the other platforms have, and then implement it with an object that is created by the engine and has a weak pointer to the engine. (If you want to do the second in a follow-up and file a bug about it that's fine, but you should do the first now since it changes the API and we may as well avoid thrashing the API when we already know what the desired API is.)